File: algorithms/java/src/test/java/leetcode/LC232ImplementQueueUsingStacks.java
package leetcode;

import org.junit.jupiter.api.Test;


import java.util.Stack;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/implement-queue-using-stacks/
 * #easy
 */
public class LC232ImplementQueueUsingStacks {
    class MyQueue {
        private Stack<Integer> stack = new Stack();
        private Stack<Integer> buffer = new Stack();

        // Push element x to the back of queue.
        public void push(int x) {
            while (stack.size() != 0) {
                buffer.push(stack.pop());
            }
            stack.push(x);
            while (buffer.size() != 0) {
                stack.push(buffer.pop());
            }
        }

        // Removes the element from in front of queue.
        public void pop() {
            stack.pop();
        }

        // Get the front element.
        public int peek() {
            return stack.peek();
        }

        // Return whether the queue is empty.
        public boolean empty() {
            return stack.size() == 0;
        }
    }

    @Test
    public void test_something() throws Exception {
        MyQueue queue = new MyQueue();
        for (int i = 0; i < 6; i++) {
            queue.push(i);
        }
        for (int i = 0; i < 6; i++) {
            assertEquals(false, queue.empty());
            assertEquals(i, queue.peek());
            queue.pop();
        }
        assertEquals(true, queue.empty());
    }
}

File: algorithms/code/leetcode/lc232_implement_queue_using_stacks/lc232_implement_queue_using_stacks.py
# https://leetcode.com/problems/implement-queue-using-stacks/

import unittest


class MyQueue:
    def __init__(self) -> None:
        self.__stack: list[int] = []
        self.__buffer: list[int] = []

    # Push element x to the back of queue.
    def push(self, x: int) -> None:
        while self.__stack:
            self.__buffer.append(self.__stack.pop())
        self.__stack.append(x)
        while self.__buffer:
            self.__stack.append(self.__buffer.pop())

    # Removes the element from in front of queue.
    def pop(self) -> int:
        return self.__stack.pop()

    # Get the front element.
    def peek(self) -> int:
        return self.__stack[-1]

    # Return whether the queue is empty.
    def empty(self) -> bool:
        return not bool(self.__stack)


class TestCode(unittest.TestCase):
    def test_something(self) -> None:
        queue = MyQueue()
        for i in range(0, 6):
            queue.push(i)
        for i in range(0, 6):
            self.assertFalse(queue.empty())
            self.assertEqual(i, queue.peek())
            queue.pop()
        self.assertTrue(queue.empty())
